File: services/google/runtimeconfig/variable.go
// Copyright 2021 Google LLC. All Rights Reserved.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package runtimeconfig

import (
	"context"
	"fmt"

	"google.golang.org/api/googleapi"
	"github.com/GoogleCloudPlatform/declarative-resource-client-library/dcl"
)

type Variable struct {
	Name          *string `json:"name"`
	RuntimeConfig *string `json:"runtimeConfig"`
	Text          *string `json:"text"`
	Value         *string `json:"value"`
	UpdateTime    *string `json:"updateTime"`
	Project       *string `json:"project"`
}

func (r *Variable) String() string {
	return dcl.SprintResource(r)
}

// Describe returns a simple description of this resource to ensure that automated tools
// can identify it.
func (r *Variable) Describe() dcl.ServiceTypeVersion {
	return dcl.ServiceTypeVersion{
		Service: "runtimeconfig",
		Type:    "Variable",
		Version: "runtimeconfig",
	}
}

const VariableMaxPage = -1

type VariableList struct {
	Items []*Variable

	nextToken string

	pageSize int32

	project string

	runtimeConfig string
}

func (l *VariableList) HasNext() bool {
	return l.nextToken != ""
}

func (l *VariableList) Next(ctx context.Context, c *Client) error {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	if !l.HasNext() {
		return fmt.Errorf("no next page")
	}
	items, token, err := c.listVariable(ctx, l.project, l.runtimeConfig, l.nextToken, l.pageSize)
	if err != nil {
		return err
	}
	l.Items = items
	l.nextToken = token
	return err
}

func (c *Client) ListVariable(ctx context.Context, project, runtimeConfig string) (*VariableList, error) {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	return c.ListVariableWithMaxResults(ctx, project, runtimeConfig, VariableMaxPage)

}

func (c *Client) ListVariableWithMaxResults(ctx context.Context, project, runtimeConfig string, pageSize int32) (*VariableList, error) {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	items, token, err := c.listVariable(ctx, project, runtimeConfig, "", pageSize)
	if err != nil {
		return nil, err
	}
	return &VariableList{
		Items:     items,
		nextToken: token,
		pageSize:  pageSize,

		project: project,

		runtimeConfig: runtimeConfig,
	}, nil
}

func (c *Client) GetVariable(ctx context.Context, r *Variable) (*Variable, error) {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	b, err := c.getVariableRaw(ctx, r)
	if err != nil {
		if dcl.IsNotFound(err) {
			return nil, &googleapi.Error{
				Code:    404,
				Message: err.Error(),
			}
		}
		return nil, err
	}
	result, err := unmarshalVariable(b, c)
	if err != nil {
		return nil, err
	}
	result.Project = r.Project
	result.RuntimeConfig = r.RuntimeConfig
	result.Name = r.Name

	c.Config.Logger.Infof("Retrieved raw result state: %v", result)
	c.Config.Logger.Infof("Canonicalizing with specified state: %v", r)
	result, err = canonicalizeVariableNewState(c, result, r)
	if err != nil {
		return nil, err
	}
	c.Config.Logger.Infof("Created result state: %v", result)

	return result, nil
}

func (c *Client) DeleteVariable(ctx context.Context, r *Variable) error {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	if r == nil {
		return fmt.Errorf("Variable resource is nil")
	}
	c.Config.Logger.Info("Deleting Variable...")
	deleteOp := deleteVariableOperation{}
	return deleteOp.do(ctx, r, c)
}

// DeleteAllVariable deletes all resources that the filter functions returns true on.
func (c *Client) DeleteAllVariable(ctx context.Context, project, runtimeConfig string, filter func(*Variable) bool) error {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	listObj, err := c.ListVariable(ctx, project, runtimeConfig)
	if err != nil {
		return err
	}

	err = c.deleteAllVariable(ctx, filter, listObj.Items)
	if err != nil {
		return err
	}
	for listObj.HasNext() {
		err = listObj.Next(ctx, c)
		if err != nil {
			return nil
		}
		err = c.deleteAllVariable(ctx, filter, listObj.Items)
		if err != nil {
			return err
		}
	}
	return nil
}

func (c *Client) ApplyVariable(ctx context.Context, rawDesired *Variable, opts ...dcl.ApplyOption) (*Variable, error) {
	c.Config.Logger.Info("Beginning ApplyVariable...")
	c.Config.Logger.Infof("User specified desired state: %v", rawDesired)

	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	// 1.1: Validation of user-specified fields in desired state.
	if err := rawDesired.validate(); err != nil {
		return nil, err
	}

	initial, desired, diffs, err := c.variableDiffsForRawDesired(ctx, rawDesired, opts...)
	if err != nil {
		return nil, fmt.Errorf("failed to create a diff: %w", err)
	}

	// TODO(magic-modules-eng): 2.2 Feasibility check (all updates are feasible so far).

	// 2.3: Lifecycle Directive Check
	var create bool
	var recreate bool
	lp := dcl.FetchLifecycleParams(opts)
	if initial == nil {
		if dcl.HasLifecycleParam(lp, dcl.BlockCreation) {
			return nil, dcl.ApplyInfeasibleError{Message: fmt.Sprintf("Creation blocked by lifecycle params: %#v.", desired)}
		}
		create = true
	} else if dcl.HasLifecycleParam(lp, dcl.BlockAcquire) {
		return nil, dcl.ApplyInfeasibleError{
			Message: fmt.Sprintf("Resource already exists - apply blocked by lifecycle params: %#v.", initial),
		}
	} else {
		for _, d := range diffs {
			if d.RequiresRecreate {
				if dcl.HasLifecycleParam(lp, dcl.BlockDestruction) || dcl.HasLifecycleParam(lp, dcl.BlockCreation) {
					return nil, dcl.ApplyInfeasibleError{
						Message: fmt.Sprintf("Infeasible update: (%v) would require recreation.", d),
					}
				}
				recreate = true
			}
			if dcl.HasLifecycleParam(lp, dcl.BlockModification) {
				return nil, dcl.ApplyInfeasibleError{Message: fmt.Sprintf("Modification blocked, diff (%v) unresolvable.", d)}
			}
		}
	}

	// 2.4 Imperative Request Planning
	var ops []variableApiOperation
	if create {
		ops = append(ops, &createVariableOperation{})
	} else if recreate {

		ops = append(ops, &deleteVariableOperation{})

		ops = append(ops, &createVariableOperation{})
		// We should re-canonicalize based on a nil existing resource.
		desired, err = canonicalizeVariableDesiredState(rawDesired, nil)
		if err != nil {
			return nil, err
		}
	} else {
		for _, d := range diffs {
			ops = append(ops, d.UpdateOp)
		}
	}
	c.Config.Logger.Infof("Created plan: %#v", ops)

	// 2.5 Request Actuation
	for _, op := range ops {
		c.Config.Logger.Infof("Performing operation %#v", op)
		if err := op.do(ctx, desired, c); err != nil {
			c.Config.Logger.Infof("Failed operation %#v: %v", op, err)
			return nil, err
		}
		c.Config.Logger.Infof("Finished operation %#v", op)
	}

	// 3.1, 3.2a Retrieval of raw new state & canonicalization with desired state
	c.Config.Logger.Info("Retrieving raw new state...")
	rawNew, err := c.GetVariable(ctx, desired.urlNormalized())
	if err != nil {
		return nil, err
	}

	// Get additional values from the first response.
	// These values should be merged into the newState above.
	if len(ops) > 0 {
		lastOp := ops[len(ops)-1]
		if o, ok := lastOp.(*createVariableOperation); ok {
			if r, hasR := o.FirstResponse(); hasR {

				c.Config.Logger.Info("Retrieving raw new state from operation...")

				fullResp, err := unmarshalMapVariable(r, c)
				if err != nil {
					return nil, err
				}

				rawNew, err = canonicalizeVariableNewState(c, rawNew, fullResp)
				if err != nil {
					return nil, err
				}
			}
		}
	}

	c.Config.Logger.Infof("Canonicalizing with raw desired state: %v", rawDesired)
	// 3.2b Canonicalization of raw new state using raw desired state
	newState, err := canonicalizeVariableNewState(c, rawNew, rawDesired)
	if err != nil {
		return nil, err
	}

	c.Config.Logger.Infof("Created canonical new state: %v", newState)
	// 3.3 Comparison of the new state and raw desired state.
	// TODO(magic-modules-eng): EVENTUALLY_CONSISTENT_UPDATE
	newDesired, err := canonicalizeVariableDesiredState(rawDesired, newState)
	if err != nil {
		return nil, err
	}
	c.Config.Logger.Infof("Diffing using canonicalized desired state: %v", newDesired)
	newDiffs, err := diffVariable(c, newDesired, newState)
	if err != nil {
		return nil, err
	}

	if len(newDiffs) == 0 {
		c.Config.Logger.Info("No diffs found. Apply was successful.")
	} else {
		c.Config.Logger.Infof("Found diffs: %v", newDiffs)
		diffMessages := make([]string, len(newDiffs))
		for i, d := range newDiffs {
			diffMessages[i] = fmt.Sprintf("%v", d)
		}
		return newState, dcl.DiffAfterApplyError{Diffs: diffMessages}
	}
	c.Config.Logger.Info("Done Apply.")
	return newState, nil
}

File: python/services/runtimeconfig/variable.py
# Copyright 2021 Google LLC. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from connector import channel
from google3.cloud.graphite.mmv2.services.google.runtimeconfig import variable_pb2
from google3.cloud.graphite.mmv2.services.google.runtimeconfig import variable_pb2_grpc

from typing import List


class Variable(object):
    def __init__(
        self,
        name: str = None,
        runtime_config: str = None,
        text: str = None,
        value: str = None,
        update_time: str = None,
        project: str = None,
        service_account_file: str = "",
    ):

        channel.initialize()
        self.name = name
        self.runtime_config = runtime_config
        self.text = text
        self.value = value
        self.project = project
        self.service_account_file = service_account_file

    def apply(self):
        stub = variable_pb2_grpc.RuntimeconfigVariableServiceStub(channel.Channel())
        request = variable_pb2.ApplyRuntimeconfigVariableRequest()
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if Primitive.to_proto(self.runtime_config):
            request.resource.runtime_config = Primitive.to_proto(self.runtime_config)

        if Primitive.to_proto(self.text):
            request.resource.text = Primitive.to_proto(self.text)

        if Primitive.to_proto(self.value):
            request.resource.value = Primitive.to_proto(self.value)

        if Primitive.to_proto(self.project):
            request.resource.project = Primitive.to_proto(self.project)

        request.service_account_file = self.service_account_file

        response = stub.ApplyRuntimeconfigVariable(request)
        self.name = Primitive.from_proto(response.name)
        self.runtime_config = Primitive.from_proto(response.runtime_config)
        self.text = Primitive.from_proto(response.text)
        self.value = Primitive.from_proto(response.value)
        self.update_time = Primitive.from_proto(response.update_time)
        self.project = Primitive.from_proto(response.project)

    def delete(self):
        stub = variable_pb2_grpc.RuntimeconfigVariableServiceStub(channel.Channel())
        request = variable_pb2.DeleteRuntimeconfigVariableRequest()
        request.service_account_file = self.service_account_file
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if Primitive.to_proto(self.runtime_config):
            request.resource.runtime_config = Primitive.to_proto(self.runtime_config)

        if Primitive.to_proto(self.text):
            request.resource.text = Primitive.to_proto(self.text)

        if Primitive.to_proto(self.value):
            request.resource.value = Primitive.to_proto(self.value)

        if Primitive.to_proto(self.project):
            request.resource.project = Primitive.to_proto(self.project)

        response = stub.DeleteRuntimeconfigVariable(request)

    @classmethod
    def list(self, project, runtimeConfig, service_account_file=""):
        stub = variable_pb2_grpc.RuntimeconfigVariableServiceStub(channel.Channel())
        request = variable_pb2.ListRuntimeconfigVariableRequest()
        request.service_account_file = service_account_file
        request.Project = project

        request.RuntimeConfig = runtimeConfig

        return stub.ListRuntimeconfigVariable(request).items

    @classmethod
    def from_any(self, any_proto):
        # Marshal any proto to regular proto.
        res_proto = variable_pb2.RuntimeconfigVariable()
        any_proto.Unpack(res_proto)

        res = Variable()
        res.name = Primitive.from_proto(res_proto.name)
        res.runtime_config = Primitive.from_proto(res_proto.runtime_config)
        res.text = Primitive.from_proto(res_proto.text)
        res.value = Primitive.from_proto(res_proto.value)
        res.update_time = Primitive.from_proto(res_proto.update_time)
        res.project = Primitive.from_proto(res_proto.project)
        return res

    def to_proto(self):
        resource = variable_pb2.RuntimeconfigVariable()
        if Primitive.to_proto(self.name):
            resource.name = Primitive.to_proto(self.name)
        if Primitive.to_proto(self.runtime_config):
            resource.runtime_config = Primitive.to_proto(self.runtime_config)
        if Primitive.to_proto(self.text):
            resource.text = Primitive.to_proto(self.text)
        if Primitive.to_proto(self.value):
            resource.value = Primitive.to_proto(self.value)
        if Primitive.to_proto(self.project):
            resource.project = Primitive.to_proto(self.project)
        return resource


class Primitive(object):
    @classmethod
    def to_proto(self, s):
        if not s:
            return ""
        return s

    @classmethod
    def from_proto(self, s):
        return s
